EngineLayer/ProteinParsimony/ProteinParsimonyEngine.cs

```diff
@@ -11,358 +12,384 @@ namespace EngineLayer
+            // and remove all peptides observed by proteins with unique peptides
+            Dictionary<string, List<Protein>> peptideSequenceToProteins = new Dictionary<string, List<Protein>>();
 
-            // *** main parsimony loop
-            bool uniquePeptidesLeft = proteinsWithUniquePeptides.Any();
+            // this dictionary associates proteins w/ all peptide sequences (list will NOT shrink over time)
+            // this is used in case of greedy algorithm ties to figure out which protein has more total peptides observed
+            Dictionary<Protein, HashSet<string>> proteinToPepSeqMatch = new Dictionary<Protein, HashSet<string>>();
 
-            int numNewSeqs = algDictionary.Max(p => p.Value.Count);
-
-            while (numNewSeqs != 0)
+            foreach (var peptide in FdrFilteredPeptides)
             {
-                var possibleBestProteinList = new List<KeyValuePair<Protein, HashSet<string>>>();
+                string sequence = peptide.BaseSequence;
+                if (TreatModPeptidesAsDifferentPeptides)
+                {
+                    //using full sequence for next three instances. might be base sequence.
+                    sequence = peptide.FullSequence;
+                }
 
-                if (uniquePeptidesLeft)
+                if (peptideSequenceToProteins.TryGetValue(sequence, out List<Protein> proteinsForThisPeptideSequence))
                 {
-                    var proteinsWithUniquePeptidesLeft = algDictionary.Where(p => proteinsWithUniquePeptides.ContainsKey(p.Key));
-                    if (proteinsWithUniquePeptidesLeft.Any())
-                    {
-                        possibleBestProteinList.Add(proteinsWithUniquePeptidesLeft.First());
-                    }
-                    else
-                    {
-                        uniquePeptidesLeft = false;
-                    }
+                    proteinsForThisPeptideSequence.Add(peptide.Protein);
+                }
+                else
+                {
+                    peptideSequenceToProteins.Add(sequence, new List<Protein> { peptide.Protein });
                 }
 
-                // gets list of proteins with the most unaccounted-for peptide base sequences
-                if (!uniquePeptidesLeft)
+                if (proteinToPepSeqMatch.TryGetValue(peptide.Protein, out var peptideSequences))
                 {
-                    possibleBestProteinList = algDictionary.Where(p => p.Value.Count == numNewSeqs).ToList();
+                    peptideSequences.Add(sequence);
                 }
+                else
+                {
+                    proteinToPepSeqMatch.Add(peptide.Protein, new HashSet<string> { sequence });
+                }
+            }
 
-                Protein bestProtein = possibleBestProteinList.First().Key;
-                HashSet<string> newSeqs = new HashSet<string>(algDictionary[bestProtein]);
+            // remove the peptides observed by proteins with unique peptides
```

Comment:
In Multiprotease Parsimony there is an bug being introduced here. I am working to find a solution.